Add port forwarding.

brendandburns opened the pull request:
@flyboarder if you get a chance to review this, definitely no obligation.

Thanks!

flyboarder commented:
@brendandburns seems reasonable, can we test for this?

brendandburns commented on examples/typescript/port-forward/port-forward.ts:
```diff
@@ -0,0 +1,16 @@
+let kc = new k8s.KubeConfig();
+kc.loadFromDefault();
+
+let forward = new k8s.PortForward(kc);
```
done

brendandburns commented on examples/typescript/port-forward/port-forward.ts:
```diff
@@ -0,0 +1,16 @@
+
+let command = process.argv[2];
+
+let kc = new k8s.KubeConfig();
```
done.

brendandburns commented on examples/typescript/port-forward/port-forward.ts:
```diff
@@ -0,0 +1,16 @@
+
+let forward = new k8s.PortForward(kc);
+
+var server = net.createServer(function(socket) {
```
done.

brendandburns commented on node-client/src/portforward.ts:
```diff
@@ -0,0 +1,50 @@
+                             output: stream.Writable | any, err: stream.Writable | any,
+                             input: stream.Readable | any): Promise<ws.connection> {
+        if (ports.length > 1) {
+            process.stderr.write('ERROR: Only one port is currently supported for port-forward');
```
done.

brendandburns commented on node-client/src/portforward.ts:
```diff
@@ -0,0 +1,50 @@
+    public async portForward(namespace: string, podName: string, ports: number[],
+                             output: stream.Writable | any, err: stream.Writable | any,
+                             input: stream.Readable | any): Promise<ws.connection> {
+        if (ports.length > 1) {
```
added.

brendandburns commented on node-client/src/portforward.ts:
```diff
@@ -0,0 +1,50 @@
+                buff = buff.slice(2);
+                firstRead[streamNum] = false;
+            }
+            // First two bytes are the port number
```
done.

brendandburns commented on node-client/src/portforward.ts:
```diff
@@ -0,0 +1,50 @@
+    }
+    // TODO: support multiple ports for real...
+    public async portForward(namespace: string, podName: string, ports: number[],
+                             output: stream.Writable | any, err: stream.Writable | any,
```
removed `any` I swear I needed it for `stdout` at some point, but this still works so ???

brendandburns commented on node-client/src/portforward.ts:
```diff
@@ -0,0 +1,50 @@
+import { WebSocketHandler } from './web-socket-handler';
+
+export class PortForward {
+    public 'handler': WebSocketHandler;
```
fixed.

brendandburns commented on node-client/src/web-socket-handler.ts:
```diff
@@ -41,10 +41,10 @@ export class WebSocketHandler {
     }
 
-    public static handleStandardInput(conn: ws.connection, stdin: stream.Readable | any) {
+    public static handleStandardInput(conn: ws.connection, stdin: stream.Readable | any, streamNum: number = 0) {
```
Retaining this b/c it will eventually be necessary for multiple port impls

brendandburns commented on node-client/src/portforward.ts:
```diff
@@ -0,0 +1,50 @@
+        this.handler = new WebSocketHandler(config);
+    }
+    // TODO: support multiple ports for real...
+    public async portForward(namespace: string, podName: string, ports: number[],
```
fixed.

brendandburns commented:
@itowlson comments mostly addressed and rebased.

Thanks!

itowlson commented on node-client/src/portforward.ts:
```diff
@@ -0,0 +1,50 @@
+        const path = `/api/v1/namespaces/${namespace}/pods/${podName}/portforward?${queryStr}`;
+        const conn = await this.handler.connect(path, null, (streamNum: number, buff: Buffer | string) => {
+            if (streamNum >= ports.length * 2) {
+                process.stdout.write(`unexpected stream: ${streamNum}\n`);
```
I still feel that writing to `process.stdout` is not appropriate here.  If it's not appropriate to throw, or to signal the problem back to the caller in some other way, then we need a way to log it that does not intrude on the experience of the _end user_ using an application that consumes this library.  Would `console.warn` to that or does that get dumped to stdout in CLI scenarios?

brendandburns commented:
@itowlson I decided to just break the connection on that error.

Added a bunch of tests too.

Split the commit into coverage commit and code commit, you can just review the code if you want.

itowlson commented:
Looks like there's a failing test though - not sure whether it's a transient networking error but it showed up on both CI runs.

brendandburns commented:
@itowlson ok, I think I fixed the tests.

Szpadel commented on node-client/src/portforward.ts:
```diff
@@ -0,0 +1,65 @@
+        } else {
+            this.handler = handler;
+        }
+        this.disconnectOnErr = true;
```
Why argument disconnectOnErr isn't used here?
